### PR TITLE
Screen Space Ambient Occlusion post effect

### PR DESCRIPTION
**Pull request by mvaligursky**
Fixes: #1136
Based on https://github.com/playcanvas/engine/pull/3092, thank you to @raytranuk for an excellent contribution. This PR in addition to the initial PR exposed few parameters and did few small changes related to parameters. I also make it to work on Webgl1.

The implementation is based on the code in Filament Engine: https://github.com/google/filament

Screenshot 2021-06-25 at 13 16 39

example of using the effect on a camera:

```
camera.addComponent("script");
camera.script.create("ssao", {
    attributes: {
        radius: 5,
        samples: 16,
        brightness: 0
    }
});
```

### CODE REVIEW

File: examples/src/examples/graphics/post-effects.tsx [hunk 1]
```diff
@@ -13,6 +13,7 @@ class PostEffectsExample extends Example {
             <AssetLoader name='bokeh' type='script' url='static/scripts/posteffects/posteffect-bokeh.js' />
             <AssetLoader name='sepia' type='script' url='static/scripts/posteffects/posteffect-sepia.js' />
             <AssetLoader name='vignette' type='script' url='static/scripts/posteffects/posteffect-vignette.js' />
+            <AssetLoader name='vignette' type='script' url='static/scripts/posteffects/posteffect-ssao.js' />
```
Comment (ellthompson) [inline examples/src/examples/graphics/post-effects.tsx:16]: Incorrect name here.

Comment (mvaligursky) [reply examples/src/examples/graphics/post-effects.tsx]: well spotted. Fixed. The name is not used in the example so it worked regardless.

File: scripts/posteffects/posteffect-ssao.js [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+        scope.resolve("uPeak2").setValue(peak * peak);
+        scope.resolve("uIntensity").setValue(intensity);
+        scope.resolve("uPower").setValue(1.0);
+        scope.resolve("uProjectionScaleRadius").setValue(projectionScale * radius);
+
+        pc.drawFullscreenQuad(device, this.targets[0], this.vertexBuffer, this.ssaoShader, rect);
+
+        scope.resolve("uSSAOBuffer").setValue(this.targets[0].colorBuffer);
+
+        // scope.resolve("uFarPlaneOverEdgeDistance").setValue(cameraFarClip / bilateralThreshold);
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:508]: Delete this line if not required?

File: scripts/posteffects/posteffect-ssao.js [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+        var colorBuffer = new pc.Texture(graphicsDevice, {
+            format: pc.PIXELFORMAT_R8_G8_B8_A8,
+            width: width,
+            height: height,
+            mipmaps: false
+        });
+        colorBuffer.minFilter = pc.FILTER_LINEAR;
+        colorBuffer.magFilter = pc.FILTER_LINEAR;
+        colorBuffer.addressU = pc.ADDRESS_CLAMP_TO_EDGE;
+        colorBuffer.addressV = pc.ADDRESS_CLAMP_TO_EDGE;
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:454]:
```suggestion
        var colorBuffer = new pc.Texture(graphicsDevice, {
            format: pc.PIXELFORMAT_R8_G8_B8_A8,
            minFilter: pc.FILTER_LINEAR,
            magFilter: pc.FILTER_LINEAR,
            addressU: pc.ADDRESS_CLAMP_TO_EDGE,
            addressV: pc.ADDRESS_CLAMP_TO_EDGE,
            width: width,
            height: height,
            mipmaps: false
        });
```

File: scripts/posteffects/posteffect-ssao.js [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+            "    gl_FragColor.a = 1.0;",
+            "}"
+        ].join("\n")
+    });
+
+    // Render targets
+    var width = graphicsDevice.width;
+    var height = graphicsDevice.height;
+    this.targets = [];
+    for (var i = 0; i < 1; i++) {
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:444]: Why is there a `for` loop here if there's only 1 iteration?

File: scripts/posteffects/posteffect-ssao.js [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+            "    float vn = dot(v, normal);  // distance * cos(v, normal)",
+            "",
+            "    // discard samples that are outside of the radius, preventing distant geometry to",
+            "    // cast shadows -- there are many functions that work and choosing one is an artistic",
+            "    // decision.",
+            "    float w = max(0.0, 1.0 - vv * uInvRadiusSquared);",
+            "    w = w*w;",
+            "",
+            "    // discard samples that are too close to the horizon to reduce shadows cast by geometry",
+            "    // not sufficently tessellated. The goal is to discard samples that form an angle 'beta'",
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:289]:
```suggestion
            " // not sufficiently tessellated. The goal is to discard samples that form an angle 'beta'",
```

File: scripts/posteffects/posteffect-ssao.js [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+
+// --------------- POST EFFECT DEFINITION --------------- //
+/**
+ * @class
+ * @name SSAOEffect
+ * @classdesc Implements the SSAOEffect post processing effect.
+ * @description Creates new instance of the post effect.
+ * @augments PostEffect
+ * @param {GraphicsDevice} graphicsDevice - The graphics device of the application.
+ * @param {any}  ssaoScript - The script using the effect.
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:12]:
```suggestion
 * @param {any} ssaoScript - The script using the effect.
```

File: scripts/posteffects/posteffect-ssao.js [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+
+    this.ssaoScript = ssaoScript;
+    this.needsDepthBuffer = true;
+
+    this.ssaoShader = new pc.Shader(graphicsDevice, {
+        attributes: {
+            aPosition: pc.SEMANTIC_POSITION
+        },
+        vshader: [
+            (graphicsDevice.webgl2) ? ("#version 300 es\n\n" + pc.shaderChunks.gles3VS) : "",
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:25]: Are we sure this line is needed? The other post effects done declare this.

File: scripts/posteffects/posteffect-ssao.js [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+            "",
+            "#define saturate(x) clamp(x,0.0,1.0)",
+            "",
+            "// Largely based on 'Dominant Light Shadowing'",
+            "// 'Lighting Technology of The Last of Us Part II' by Hawar Doghramachi, Naughty Dog, LLC",
+            "",
+            "const float kSSCTLog2LodRate = 3.0;",
+            "",
+            "struct ConeTraceSetup {",
+            "   // fragment info",
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:56]:
```suggestion
            "    // fragment info",
```

File: scripts/posteffects/posteffect-ssao.js [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+            "    highp vec2 ssStartPos;",
+            "    highp vec3 vsStartPos;",
+            "    vec3 vsNormal;",
+            "",
+            "   // light (cone) info",
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:61]:
```suggestion
            "    // light (cone) info",
```

File: scripts/posteffects/posteffect-ssao.js [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+            "    vec3 vsConeDirection;",
+            "    float shadowDistance;",
+            "    float coneAngleTangeant;",
+            "    float contactDistanceMaxInv;",
+            "    vec2 jitterOffset;          // (x = direction offset, y = step offset)",
+            "",
+            "   // scene infos",
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:68]:
```suggestion
            "    // scene infos",
```

File: scripts/posteffects/posteffect-ssao.js [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+            "    highp mat4 screenFromViewMatrix;",
+            "    float projectionScale;",
+            "    vec4 resolution;",
+            "    float maxLevel;",
+            "",
+            "   // artistic/quality parameters",
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:74]:
```suggestion
            "    // artistic/quality parameters",
```

File: scripts/posteffects/posteffect-ssao.js [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+            "varying vec2 vUv0;",
+            "",
+            "void main(void)",
+            "{",
+            "    gl_Position = vec4(aPosition, 0.0, 1.0);",
+            "    vUv0 = (aPosition.xy + 1.0) * 0.5;",
+            "}"
+        ].join("\n"),
+        fshader: [
+            (graphicsDevice.webgl2) ? ("#version 300 es\n\n" + pc.shaderChunks.gles3PS) : "",
```
Comment (willeastcott) [inline scripts/posteffects/posteffect-ssao.js:37]: Again, just checking this line is required.

**Review by willeastcott — APPROVED** [review]
Approving with various suggestions and comments.

File: scripts/posteffects/posteffect-ssao.js (hunk 4 above)
Comment (mvaligursky) [reply scripts/posteffects/posteffect-ssao.js:444]: copy & paste from bloom, but changing 2 to 1 iterations. I'll clean this up.

File: scripts/posteffects/posteffect-ssao.js (hunk 7 above)
Comment (mvaligursky) [reply scripts/posteffects/posteffect-ssao.js:25]: yes those are neeed to use derivatives in webgl2. Chunk based shaders handle this automatically, but manual shaders don't.
But while looking at this, I got SSAO to work on webgl1 as well .. so worth spending time on this!

**Comment by Maksims** [comment]
Can please the shader code be moved to a separate file? Making it more usable than it is when written within JS?
Editor users don't have the easy ability to copy it.

**Comment by mvaligursky** [comment]
I agree it's not ideal, but it matches the style of all post effects - it's a single file with all shaders inlined.
But definitely something to consider for the future, we have plans for the next generation of post effects (including node graph support and similar).

**Pull request merged**